### DISCUSSION TITLE
Recommendation: Write plug-ins using TypeScript

### DISCUSSION THREAD

**Discussion by Gk0Wk** [SmilyOrg/tw5-monaco#9] [Ideas]
I think larger projects should switch to using TypeScript instead of JavaScript, which helps reduce the cost of maintaining the code and the difficulty of writing it. This has the added benefit that I and others have implemented TiddlyWiki5's TypeScript definition: [TW5-Typed](https://github.com/tiddly-gittly/TW5-Typed), which can make development easier.

[Here](https://github.com/tiddly-gittly/TiddlyWiki-TS-Plugin-Template) is a template project for developing TW plugins based on TypeScript, supporting both GitHub Action builds as well as local builds (using some of my previous work, such as [tw5-plugin-packer](https://github.com/tiddly-gittly/tw5-plugin-packer), to implement GitHub Action builds and local builds of plugins).

**Comment by Gk0Wk** [comment]
Then, of course, can package Monaco into an offline version.

**Comment by SmilyOrg** [comment]
TypeScript is great, however I'd like to avoid huge or slow compile steps (e.g. webpack) if possible. I'll have to check out TW5-Typed and the other links in a bit, thanks!

**Comment by Gk0Wk** [comment]
I use Rollup to package CodeMirror-Enhanced and it takes about a few seconds to finish the build. So I think it can be done fast enough using rollup.

Further, you can also use [esbuild](https://esbuild.github.io/), which is claimed to be the fastest TS compiler ever, although there are still some contents that cannot be packaged with esbuild, I think it's enough just for TW plugins.

If you think it's ok, next week I'll try to transform the monaco plugin into a ts project built using esbuild or rollup.

**Reply by SmilyOrg** [reply]
TW5-Typed is indeed a cool project! I keep having to search for functions in TW5 source to know what's even available and how to call it 😓

But I'm wondering if bundling is even necessary? [Maybe not?](https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755)

Could we also just use `tsc` / `tsc --watch` to convert `.ts` to `.js` during build/dev instead of using esbuild/rollup?

Would be interesting to see what it would look like with TypeScript, I definitely started with JS mostly because it was easier to set up at first.
